### ISSUE TITLE
Base path to reverse proxy

### ISSUE THREAD

**Issue by fana13** [alexjustesen/speedtest-tracker#313]
It will be great if we could specify a base path through the docker environment, like this : -e BASE_PATH=/speedtest

**Comment by alexdelprete** [comment]
If you search the issues, there are many discussions on reverse proxies setup, Traefik etc.

**Comment by fana13** [comment]
I've read many discusssions about this, but not really that issue.
It's about subfolder nginx for reverse proxy.
APP_URL in .env it's not for that, right ?

**Comment by alexdelprete** [comment]
the app is listening on 80 and 443, you should be able to address this with NGINX without requiring modifications.

What exactly are you trying to achieve?

**Comment by fana13** [comment]
I try to have https://myowndomain/speedtracker

**Comment by alexdelprete** [comment]
With Traefik I can do that using a prefix after the domain. I guess it's doable also with NGINX if that's what you use, you don't need basepath on the app.

**Comment by Racle** [comment]
> With Traefik I can do that using a prefix after the domain. I guess it's doable also with NGINX if that's what you use, you don't need basepath on the app.

How you do this in Traefik?

When using with nginx `/speedtracker` path shows 404 and docker logs show this:
` "GET /speedtracker/ HTTP/1.0" 404 6603 "`


If I put this to nginx (basically rewriting /speedtracker/ to /)
`rewrite /speedtracker/(.*) /$1 break;`

It will try to redirect to mydomain.com/admin after that, which is also incorrect behaviour as it should redirect to mydomain.com/speedtracker/admin.

With older speedtest-tracker using environment key `APP_URL=/speedtracker` this worked perfectly.

I also tried to use same env config with this one without success. Same thing when trying to use `APP_URL=mydomain.com/speedtracker`.

**Comment by fana13** [comment]
Same for me, it doesn't work.
For example, with the older henrywhitaker's repo, this nginx configuration work like a charm.
@alexdelprete could you unclose this issue, it seem real

**Comment by alexdelprete** [comment]
> how you do this in Traefik?

https://doc.traefik.io/traefik/middlewares/http/stripprefix/

![image](https://user-images.githubusercontent.com/7027842/212777921-edff437e-bfd7-486d-bcce-f89ed9129c84.png)

**Comment by alexdelprete** [comment]
> Same for me, it doesn't work.

what exactly doesn't work? Speedtest-Tracker is not a proxy, nor a webserver, it's a webapp.

you can say something doesn't work when its documentation describes a feature, a configuration, and that doesn't work. this is not the case. your expectations about an app is not a bug.

this webapp is listening on http/https ports, no path/subpaths, but on root. it's up to your reverse proxy to be configured like you want it to behave, in order to proxy correctly this webapp on root path. On Traefik StripPrefix can be used, on NGINX I have no idea because I never used it. Check NGINX documentation on how to achieve this and don't blame it on the webapp.

Furthermore, I'm not the maintainer/dev, that would be @alexjustesen and he already closed the case because it's not a bug, it's something the reverse proxy has to manage.

**Comment by Racle** [comment]
> > how you do this in Traefik?
> 
> https://doc.traefik.io/traefik/middlewares/http/stripprefix/
> 
> ![image](https://user-images.githubusercontent.com/7027842/212777921-edff437e-bfd7-486d-bcce-f89ed9129c84.png)

This doesn't work, same redirect problem with traefik when using 
![image](https://user-images.githubusercontent.com/759091/212826698-805a78eb-67e8-4ae5-a585-b61a005884e6.png)


It tries to redirect `mydomain/admin` when it should be redirecting `mydomain/speedtest-tracker/admin`

Base url has to be programmed into application, middleware/reverse proxy cannot fix this as far as I know.

If you have working configuration, feel free to share.

**Comment by fana13** [comment]
> > Same for me, it doesn't work.
> 
> what exactly doesn't work? Speedtest-Tracker is not a proxy, nor a webserver, it's a webapp.
> 
> you can say something doesn't work when its documentation describes a feature, a configuration, and that doesn't work. this is not the case. your expectations about an app is not a bug.
> 
> this webapp is listening on http/https ports, no path/subpaths, but on root. it's up to your reverse proxy to be configured like you want it to behave, in order to proxy correctly this webapp on root path. On Traefik StripPrefix can be used, on NGINX I have no idea because I never used it. Check NGINX documentation on how to achieve this and don't blame it on the webapp.
> 
> Furthermore, I'm not the maintainer/dev, that would be @alexjustesen and he already closed the case because it's not a bug, it's something the reverse proxy has to manage.

I don't tag it as a bug, tag was FEATURE and i think it's a real feature.
So if you can't have any idea because you never use it, why did you close this ?
I have checked NGINX documentation, i have a lot of app in my nginx configuration.
I gave you an example with the older repo, which have added a APP_PATH or PATH_URL in the .env, that's why i asked if this feature will be implemented, nothing more

**Comment by alexdelprete** [comment]
> I don't tag it as a bug, tag was FEATURE and i think it's a real feature.

Sorry, I tought you opened it as an issue, my bad.

> So if you can't have any idea because you never use it, why did you close this ?

I'm not the maintainer, you are confusing names.

> I gave you an example with the older repo

This is not the old project, it's a new one. You can achieve what you want if you configure properly your reverse proxy, if your reverse proxy doesn't allow to do it, change proxy.

**Comment by alexdelprete** [comment]
> Base url has to be programmed into application, middleware/reverse proxy cannot fix this as far as I know.

Quote from traefik docs:

> Use a StripPrefix middleware if your backend listens on the root path (/) but should be exposed on a specific prefix.

Feel free to look for working configs in Traefik community forums, there are a lot of threads, here's one: https://community.traefik.io/t/traefik-v2-6-pathprefix-wont-work/13398/2

**Comment by Racle** [comment]
> Use a StripPrefix middleware if your backend listens on the root path (/) but should be exposed on a specific prefix.

@alexdelprete and that's the issue, this applies to `backend`.  I see how you would think this would work in situation like this when it doesn't

If I had an API which calls `mydomain/api`, I could redirect that with `mydomain/speedtracker/api` and it would work just fine as backend _usually_ doesn't try to redirect user to another address.

When connecting to speedtest-tracker frontend tries to redirect to `/admin` path and you cannot change how that works with middleware, you have to change that in code.

It also hardcodes assets path to `/`, so even when I could load page without redirecting, assets aren't loading correctly.

And even on your stackoverflow link it's stated that it will only work if your application works like that:

> This will only work as long as the app you are trying to start does NOT set a custom html <base> to something hardcoded like root /.


Traefik isn't the solution here. Code changes are required to set correct base path.

**Comment by fana13** [comment]
@alexdelprete i'm sorry, i misconfusing yours nicks
@alexjustesen could you unclose this feature demand ?
Thanks both for your time

**Comment by alexdelprete** [comment]
> and that's the issue, this applies to `backend`. I see how you would think this would work in situation like this when it doesn't

It does...and that's the only way to make it work with Laravel, Laravel still doesn't officially support serving an app through subdirs.

![image](https://user-images.githubusercontent.com/7027842/213251347-f432a80e-d436-4940-be23-dfe1e9361575.png)

Luckily they merged this [PR](https://github.com/laravel/framework/pull/40014), which leverages X-Forwarded-Prefix. And Traefik's StripPrefix also creates that header exactly for these use-cases.

> When connecting to speedtest-tracker frontend tries to redirect to `/admin` path and you cannot change how that works with middleware, you have to change that in code.

Sure you can, because Traefik sets `X-Forwarded-Prefix` header, and the backend (Laravel in this specific case) should use that to build the relative URLs, prepending the Prefix:

![image](https://user-images.githubusercontent.com/7027842/212964989-4f75bce2-a63b-431c-93e4-f850f2a3417d.png)

Now I found out that the configuration of Laravel in Speedtest-Tracker didn't include the support of X-Forwarded-Prefix, so I'll try to include it and do some tests.

**Comment by alexdelprete** [comment]
> i'm sorry, i misconfusing yours nicks

our names...these are our real names, not nicknames. :)



> @alexjustesen could you unclose this feature demand ?
> Thanks both for your time

check my reply above to Racle, with some technical details.

**Comment by alexjustesen** [comment]
@alexdelprete the bulk of the questions you seem to target are proxy related, espcially when dealing with Traefix. Would you be up for writing a section in the docs in extreme detail on how to setup reverse proxies and then how they can be used with Speedtest Tracker? Thinking we can eliminate a lot of these questions and issues if we get some good docs written.

If you're interested we can connect on Telegram or Discord.

**Comment by alexdelprete** [comment]
> Would you be up for writing a section in the docs in extreme detail on how to setup reverse proxies and then how they can be used with Speedtest Tracker? Thinking we can eliminate a lot of these questions and issues if we get some good docs written.

Sure, I'd love to contribute, but first I need to address some issues in order to find the proper configuration of the project. Let's connect on Discord: `@alexdelprete#5566`.

**Comment by vitormfgoncalves** [comment]
> > Base url has to be programmed into application, middleware/reverse proxy cannot fix this as far as I know.
> 
> Quote from traefik docs:
> 
> > Use a StripPrefix middleware if your backend listens on the root path (/) but should be exposed on a specific prefix.
> 
> That is exactly what you're trying to achieve, that's the use case.
> 
> Feel free to look for working configs in Traefik community forums, there are a lot of threads, here's one: https://community.traefik.io/t/traefik-v2-6-pathprefix-wont-work/13398/2
> 
> Also:
> 
> * https://community.traefik.io/t/pathprefix-and-strip/2122/3?u=alexdelprete
> * https://community.traefik.io/t/traefik-is-unable-to-strip-prefix-before-forwarding-request-to-a-
> * https://stackoverflow.com/questions/73680522/traefik-proxypass-using-stripprefix-i-get-a-404-for-sub-queries
> * https://medium.com/predictly-on-tech/using-traefik-in-docker-compose-abc1c327d7f7
> * https://www.reddit.com/r/Traefik/comments/lss1ie/cannot_make_stripprefix_work_help/
> 
> I don't use this config, I use `speedtest.mydomain.dom`.

I don't understand why you refuse to accept a valid argument and a true problem.
Not everyone uses traefik, and reverse proxies are not magic.
Most of the web runs on Javascript (webapps included). When you reverse-proxy a webapp, the JS still contains the incorrect paths. Even if you rewrite the server's response (aka the javascript), which I think you are referring to some posts above, it's almost certain that you will run into trouble.
Finally, not everyone can use a subdomain. I can't (at home).

**Comment by alexdelprete** [comment]
> I don't understand why you refuse to accept a valid argument and a true problem.

Because you don't read carefully, or you didn't understand what you read, so I'll be even more clear:

1. Laravel does NOT support subfolders. Full stop. They merged only recently a PR that allows to circumvent the issue, and that's what I'm thinking to follow in order to make it work.
2. In order to make it work, Laravel (Simfony actually) leverages specific headers to build the path of resources, so when you say "the JS still contains the incorrect path" it's clear you don't understand what we're talking about. I'm not rewriting the server's response, it's Laravel through headers building paths dynamically. And this works not only for Traefik but with all reverse proxies that support X-Forwarded-Prefix. NGINX supports it, and I bet others too.
3. Finally, I didn't say USE SUBDOMAINS, I said: "I use subdomains", I hope you understand the difference. Furthermore, Laravel devs official reply is: "use subdomains, subfolders ARE NOT SUPPORTED".

Got it now? Hope so.

**Comment by alexdelprete** [comment]
> Any news on this?

X-Forwarded-Prefix has been implemented in Januray: PR #342

**Comment by alpe12** [comment]
> > Any news on this?
> 
> `X-Forwarded-Prefix` has been implemented in January: PR #342

For me it still doesn't work correctly.
Most urls works. But all urls under /livewire remains at / instead of the configured prefix.

Nginx.
```
location ^~ /speedtest-tracker/ {
    proxy_pass http://speedtest-tracker:80/;
    proxy_set_header Host "localhost";
    proxy_set_header X-Forwarded-Prefix /speedtest-tracker;
    proxy_set_header X-Forwarded-Proto $scheme;
    access_log off;
}
```
